### PR TITLE
issue: 1008712 Synchronize access to intermediate rx buffer for bonding

### PR DESCRIPTION
**Pull request by igor-ivanov**
m_buffer_per_ring is introduced to avoid buffer allocation
in fast path but it should be clean before usage.

Signed-off-by: Igor Ivanov <igor.ivanov.va@gmail.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1065/ for details (Mellanox internal link).

File: src/vma/dev/ring_bond.cpp [hunk 1]
```diff
@@ -394,6 +394,7 @@ bool ring_bond::reclaim_recv_buffers(descq_t *rx_reuse)
 {
 	devide_buffers_helper(rx_reuse, m_buffer_per_ring);
 	for (uint32_t i = 0; i < m_n_num_resources; i++) {
+		m_buffer_per_ring[i].clear_without_cleanup();
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: clear_without_cleanup() will make size() to return 0. Do we expect the next if statement to succeed?
```
if (m_buffer_per_ring[i].size() > 0) {
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1068/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1069/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1081/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1643/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1647/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1085/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1792/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1240/ for details (Mellanox internal link).

**Comment by liranoz12** [comment]
@OphirMunk, @igor-ivanov,

The solution seems to be correct but i dont know how it will effect in terms of performance.
Each _ring_bond::reclaim_recv_buffers_ will create array of 10 list objects while the common amount of slaves for bond will be probably two.
What about adding a new lock which will only be used at _ring_bond::reclaim_recv_buffers_ ?
[0001-issue-1008712-Synchronize-access-to-intermediate-rx-patch.txt](https://github.com/Mellanox/libvma/files/1010909/0001-issue-1008712-Synchronize-access-to-intermediate-rx-patch.txt)

Because the common amount of slaves for a bond is two this will be handled very quickly.
Additionally, there are more locks inside the function : ring_simple lock, buffer_pool lock so adding another lock should not affect too much.

Liran.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1793/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1241/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1794/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1242/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1798/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1246/ for details (Mellanox internal link).

File: src/vma/dev/ring_bond.cpp [hunk 2]
```diff
@@ -53,15 +53,21 @@
...
 #define MODULE_HDR	 	MODULE_NAME "%d:%s() "
 
+/* Set limitation for number of rings for bonding device */
+#define MAX_NUM_RING_RESOURCES 10
+
+
 ring_bond::ring_bond(int count, net_device_val::bond_type type, net_device_val::bond_xmit_hash_policy bond_xmit_hash_policy, uint32_t mtu) :
 ring(count, mtu), m_lock_ring_rx("ring_bond:lock_rx"), m_lock_ring_tx("ring_bond:lock_tx") {
+	if (m_n_num_resources > MAX_NUM_RING_RESOURCES) {
+		ring_logpanic("Error creating bond ring with more than %d resource", MAX_NUM_RING_RESOURCES);
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: @igor-ivanov logpanic should only be used for unexpected SW flows. In this case the number of rings per bond is user configuration parameter and as such we should only alert the user, but not crashing the system.

File: src/vma/dev/ring_bond.cpp [hunk 3]
```diff
@@ -392,17 +395,24 @@ void ring_bond::inc_tx_retransmissions(ring_user_id_t id)
 
 bool ring_bond::reclaim_recv_buffers(descq_t *rx_reuse)
 {
-	devide_buffers_helper(rx_reuse, m_buffer_per_ring);
+	/* use this local array to avoid locking mechanizm
+	 * for threads synchronization. So every thread should use
+	 * own array. Set hardcoded number to meet C++11
+	 * VLA is not an official part of C++11.
+	 */
+	descq_t buffer_per_ring[MAX_NUM_RING_RESOURCES];
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: In the  original code m_buffer_per_ring was created as:
```
m_buffer_per_ring = new descq_t[m_n_num_resources + 1];
```
while in your code you are creating buffer_per_ring array size smaller by 1:
```
descq_t buffer_per_ring[MAX_NUM_RING_RESOURCES];
```

Please explain why.

File: src/vma/dev/ring_bond.cpp [hunk 4]
```diff
@@ -392,17 +395,24 @@ void ring_bond::inc_tx_retransmissions(ring_user_id_t id)
...
+			if (!m_bond_rings[i]->reclaim_recv_buffers(&buffer_per_ring[i])) {
+				g_buffer_pool_rx->put_buffers_after_deref_thread_safe(&buffer_per_ring[i]);
 			}
 		}
 	}
 
-	if (m_buffer_per_ring[m_n_num_resources].size() > 0)
-		g_buffer_pool_rx->put_buffers_after_deref_thread_safe(&m_buffer_per_ring[m_n_num_resources]);
+	if (buffer_per_ring[m_n_num_resources].size() > 0)
+		g_buffer_pool_rx->put_buffers_after_deref_thread_safe(&buffer_per_ring[m_n_num_resources]);
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: Here you may have array overflow when accessing:
```
buffer_per_ring[m_n_num_resources]
```
if m_n_num_resources is 10 you are out of range

**Comment by OphirMunk** [comment]
@igor this PR will be merged to fix the crash. Please open a new PR to address the review comments

**Comment by OphirMunk** [comment]
Merged #354

**Pull request closed without merging**